java-checks/src/main/java/org/sonar/java/checks/VirtualThreadUnsupportedMethodsCheck.java

```diff
@@ -0,0 +1,99 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.JavaVersionAwareVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6901")
+public class VirtualThreadUnsupportedMethodsCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  private static final String ISSUE_MESSAGE = "Method '%s' is not supported on virtual threads.";
+
+  private static final MethodMatchers VIRTUAL_THREAD_BUILDER_METHODS = MethodMatchers.or(
+    MethodMatchers.create()
+      .ofTypes("java.lang.Thread$Builder$OfVirtual")
+      .names("unstarted", "start")
+      .addParametersMatcher("java.lang.Runnable").build(),
+    MethodMatchers.create()
+      .ofTypes("java.lang.Thread")
+      .names("startVirtualThread")
+      .addParametersMatcher("java.lang.Runnable").build()
+  );
+
+  private static final MethodMatchers VIRTUAL_THREAD_UNSUPPORTED_METHODS = MethodMatchers.create()
+    .ofTypes("java.lang.Thread")
+    .names("setDaemon", "setPriority", "getThreadGroup")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava21Compatible();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree mit = (MethodInvocationTree) tree;
+    if (VIRTUAL_THREAD_UNSUPPORTED_METHODS.matches(mit)) {
+      checkMethodCalledOnVirtualThread(mit);
+    }
+  }
+
+  private void checkMethodCalledOnVirtualThread(MethodInvocationTree methodInvocation) {
+    var memberSelect = (MemberSelectExpressionTree) methodInvocation.methodSelect();
+    var expression = memberSelect.expression();
+    if (isIdentifierAndVirtualThread(expression) || isMethodInvocationAndReturningVirtualThread(expression)) {
```

Comment:
I would change `isMethodInvocationAndReturningVirtualThread(expression)` to return nullable or Optional `MethodInvocationTree` evidence about virtual threads instead of a boolean. 
Then I would add `ExpressionUtils.methodName(virtualThreadMethodInvocation)` as a secondary location.